src/Stack/Setup.hs

```diff
@@ -235,7 +235,7 @@ setupEnv mResolveMissingGHC = do
             , soptsGHCJSBootOpts = ["--clean"]
             }
 
-    (mghcBin, compilerBuild, _) <- ensureCompiler sopts
+    (mghcBin, compilerBuild, _) <- if bcClean bconfig then pure (Nothing, CompilerBuildStandard, False) else ensureCompiler sopts
```

Comment:
This change is making the presence of a compiler optional, and so the code should reflect this — yet this implementation does not.

Subsequent code uses the `compilerBuild` option; it is faked here, making that code more fragile. It might be better that it has type `Maybe CompilerBuild` here; in this way the possible absence of a compiler is reflected in the type system. This then propagates to `EnvConfig` which I believe will need changes.